### ISSUE TITLE
Building ccls on Ubuntu 16.04 with clang binaries

### ISSUE THREAD

**Issue by aste4** [MaskRay/ccls#469]
I've spent a lot of time trying to build this on Ubuntu 16. Because I don't want install GCC 7, I download the pre-built binaries of clang 7 or clang 8. I install it in /usr/local/ and set -DCMAKE_CXX_COMPILER=clang++. The problem is that clang++ tries to use gcc stdlib. I've found success compiling by using -DCMAKE_CXX_FLAG to tell clang to compile with clang stdlib but it fails during linking.

I then found -DCLANG_USE_BUNDLED_LIBC in the documentation but that option is not valid anymore since it was removed in https://github.com/MaskRay/ccls/commit/9bc762961a9f7fac5e92c2169801b6e2c6f3614f

Please help me build this on Ubuntu 16 using clang and not gcc.

**Comment by MaskRay** [comment]
On Linux, clang defaults to use libstdc++.

```cpp
ToolChain::CXXStdlibType Linux::GetDefaultCXXStdlibType() const {
  if (getTriple().isAndroid())
    return ToolChain::CST_Libcxx;
  return ToolChain::CST_Libstdcxx;
}
```

If the prebuilt archive of clang 7 or clang 8 was built with libstdc++, you have to provide libstdc++.so when linking ccls. The removed `-DCLANG_USE_BUNDLED_LIBC++` will not help you.

**Comment by unphased** [comment]
@aste4 I can confirm on 16.04 here that upon doing the first step of installing g++-7 from packages (but without running `update-alternatives` to override the system default compiler), and invoking cmake with 

```
cmake -H. -BRelease -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=/home/slu/Downloads/clang+llvm-7.0.1-x86_64-linux-gnu-ubuntu-16.04/ -DCMAKE_CXX_COMPILER=/usr/bin/g++-7
```

worked for me to succeed building ccls.

**Comment by aste4** [comment]
@unphased I think I ended up doing that as well.

Thanks!

**Comment by unphased** [comment]
This built ccls behaves differently from the ccls that I later obtained through a snap. 

Neither are behaving as expected, working half the time and mysteriously failing to resolve symbols that are clearly indicated by :cocInfo to have their headers loaded from the compile_commands.json... 

I am researching additional troubleshooting/sanity checking features in ccls, but I'm getting close to giving up on it. The issue I'm having is that the correct headers are shown to be loaded.

**Comment by aste4** [comment]
Is this with C or C++? C in my experience works ok. with C++ on Debian Buster, the go-to-definition for headers tries to open `/include/c++/8/iostream` which is just an empty file. There's probably a way to fix that with .ccls but it has not annoyed me enough.

**Comment by unphased** [comment]
I ran `jq -s add  $(find -name compile_commands.json) > compile_commands.json` from my workspace directory to simply concatenate all contents of individual CDBs generated by cmake in my codebase into a massive single `compile_commands.json` CDB file. 

Subsequently starting vim on a cpp file saw the `ccls` process crunch for about 10 minutes consuming all CPU, which made me very happy as it was finally indexing away happily. 

At this point everything seems to be working fine for me. What I was missing until now was a combination of not understanding that I should wait for the indexing to complete and could monitor that with top, and also it appears that by combining the CDB content manually fills in all the holes. I still don't understand the strange behavior from before (as before I used only the CDB from one build artifact, but I still saw strange problems where ccls couldn't figure out basic things even when I browsed source files related to that specific build artifact), but I'm really glad my efforts have finally paid off.
